#Library

In [ ]:
!pip install shap
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)

import pandas as pd
import numpy as np
import seaborn as sns
import re
import matplotlib.pyplot as plt
import shap

from imblearn import under_sampling, over_sampling
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import ConfusionMatrixDisplay
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import learning_curve, GridSearchCV
from sklearn.linear_model import LogisticRegression, RidgeClassifier, BayesianRidge, LinearRegression
from sklearn.metrics import classification_report,ConfusionMatrixDisplay
from sklearn.metrics import accuracy_score, recall_score, precision_score
from sklearn.metrics import confusion_matrix, roc_auc_score, f1_score, recall_score
from sklearn import svm
from xgboost import XGBClassifier
sns.set()
%matplotlib inline

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 540.1/540.1 kB 18.0 MB/s eta 0:00:00


#DataFrame

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
%cd /content/drive/MyDrive/DatasetFrame/Spam Email Dataset

/content/drive/MyDrive/DatasetFrame/Spam Email Dataset


In [ ]:
df = pd.read_csv('spam_email_dataset.csv', encoding='latin')
df.head(5)

Email  \
0     mikerusso@example.net   
1   waynebailey@example.org   
2        jill43@example.com   
3  johnsonkaren@example.org   
4    markwilson@example.org   

                                             Subject  \
0                       Even hotel community church.   
1  Try themselves guess fight white agreement thu...   
2  Environmental commercial off seem any conference.   
3      Smile real TV father commercial day increase.   
4                     Fast stage he oil institution.   

                     Sender                   Recipient        Date   Time  \
0    emilyscott@example.org    gregorysmith@example.org  13-02-2023  04:13   
1      annwhite@example.net  gonzalezdaniel@example.net  09-08-2023  06:15   
2       david88@example.net   michellebaker@example.net  16-05-2023  01:32   
3  lindaalvarez@example.com   schroedertodd@example.com  25-04-2023  14:50   
4     vstafford@example.com   emilywilliams@example.com  11-07-2023  21:44   

   Attachments  Link Count  Word Count  Uppercase Count  Exclamation Count  \
0            3           0         191               32                  3   
1            3           9          45                1                  2   
2            3           9          52                1                  4   
3            2           3          75                6                  0   
4            3           5         299              279                  0   

   Question Count  Dollar Count  Punctuation Count  HTML Tags Count  \
0               3             0                 11                1   
1               1             2                  8                2   
2               0             2                 10                1   
3               2             0                  8                2   
4               2             0                  6                1   

   Spam Indicator  
0               1  
1               1  
2               1  
3               0  
4               1

#DataFrame Info

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6000 entries, 0 to 5999
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Email              6000 non-null   object
 1   Subject            6000 non-null   object
 2   Sender             6000 non-null   object
 3   Recipient          6000 non-null   object
 4   Date               6000 non-null   object
 5   Time               6000 non-null   object
 6   Attachments        6000 non-null   int64 
 7   Link Count         6000 non-null   int64 
 8   Word Count         6000 non-null   int64 
 9   Uppercase Count    6000 non-null   int64 
 10  Exclamation Count  6000 non-null   int64 
 11  Question Count     6000 non-null   int64 
 12  Dollar Count       6000 non-null   int64 
 13  Punctuation Count  6000 non-null   int64 
 14  HTML Tags Count    6000 non-null   int64 
 15  Spam Indicator     6000 non-null   int64 
dtypes: int64(10), object(6)
memory usage: 750.

In [ ]:
df.describe()

Attachments   Link Count   Word Count  Uppercase Count  \
count  6000.000000  6000.000000  6000.000000      6000.000000   
mean      1.497833     5.005500   153.269500        75.997667   
std       1.125555     3.166092    83.135747        64.735568   
min       0.000000     0.000000    10.000000         0.000000   
25%       0.000000     2.000000    83.000000        23.000000   
50%       1.000000     5.000000   154.000000        59.000000   
75%       3.000000     8.000000   223.000000       114.000000   
max       3.000000    10.000000   300.000000       300.000000   

       Exclamation Count  Question Count  Dollar Count  Punctuation Count  \
count        6000.000000     6000.000000   6000.000000        6000.000000   
mean            2.481833        1.486000      0.989500           7.549667   
std             1.717895        1.118785      0.812301           4.624619   
min             0.000000        0.000000      0.000000           0.000000   
25%             1.000000        0.000000      0.000000           4.000000   
50%             2.000000        1.000000      1.000000           8.000000   
75%             4.000000        2.000000      2.000000          12.000000   
max             5.000000        3.000000      2.000000          15.000000   

       HTML Tags Count  Spam Indicator  
count      6000.000000     6000.000000  
mean          2.528833        0.497000  
std           1.703522        0.500033  
min           0.000000        0.000000  
25%           1.000000        0.000000  
50%           3.000000        0.000000  
75%           4.000000        1.000000  
max           5.000000        1.000000

## Unique Value Check

In [ ]:
def print_unique_value(df):
    for col in df.columns:
        print(f"Column: {col}")
        print(df[col].unique())
        print("="*10)
print_unique_value(df)

Column: Email
['mikerusso@example.net' 'waynebailey@example.org' 'jill43@example.com'
 ... 'cassandravazquez@example.org' 'amoore@example.com'
 'vwilliams@example.org']
Column: Subject
['Even hotel community church.'
 'Try themselves guess fight white agreement thus although.'
 'Environmental commercial off seem any conference.' ...
 'Matter issue win agent likely.' 'Again member father.'
 'Position recognize behind boy try able.']
Column: Sender
['emilyscott@example.org' 'annwhite@example.net' 'david88@example.net' ...
 'dlewis@example.org' 'eperry@example.org' 'jason06@example.org']
Column: Recipient
['gregorysmith@example.org' 'gonzalezdaniel@example.net'
 'michellebaker@example.net' ... 'cervanteskimberly@example.org'
 'dana02@example.org' 'sjones@example.com']
Column: Date
['13-02-2023' '09-08-2023' '16-05-2023' '25-04-2023' '11-07-2023'
 '13-01-2023' '18-01-2023' '02-05-2023' '27-02-2023' '21-06-2023'
 '24-04-2023' '04-06-2023' '19-07-2023' '13-08-2023' '09-01-2023'
 '17-07-2023'

##Missing Value Info

In [ ]:
def find_missing_value(data_frame):
    missing_data = data_frame.isnull().sum(axis=0).reset_index()
    missing_data.columns = ['variable', 'missing values']
    missing_data['filling factor (%)'] = (data_frame.shape[0]-missing_data['missing values'])/data_frame.shape[0]*100
    return missing_data.sort_values('filling factor (%)').reset_index(drop = True)

find_missing_value(df)

variable  missing values  filling factor (%)
0               Email               0               100.0
1             Subject               0               100.0
2              Sender               0               100.0
3           Recipient               0               100.0
4                Date               0               100.0
5                Time               0               100.0
6         Attachments               0               100.0
7          Link Count               0               100.0
8          Word Count               0               100.0
9     Uppercase Count               0               100.0
10  Exclamation Count               0               100.0
11     Question Count               0               100.0
12       Dollar Count               0               100.0
13  Punctuation Count               0               100.0
14    HTML Tags Count               0               100.0
15     Spam Indicator               0               100.0

The data at 100% Fill

##Duplicate Check

In [ ]:
duplicates = df[df.duplicated(keep=False)]
print(duplicates)

Empty DataFrame
Columns: [Email, Subject, Sender, Recipient, Date, Time, Attachments, Link Count, Word Count, Uppercase Count, Exclamation Count, Question Count, Dollar Count, Punctuation Count, HTML Tags Count, Spam Indicator]
Index: []


# Data Info



*   Dari DataFrame terdapat 6000 data dan 16 kolom
*   `Spam Indicator` (1) is Spam (0) Non Spam
*   Target kolom `Spam Indicator` menjelaskan bahwa 50% data adalah NonSpam (1)
*   Dari Mean `Link Count` up to 5 kemungkinan Spam 49%
*   sama dengen `Uppercase Count` 75% likely sebagai Spam 49%
*   `HTML Tags Count` pada 2.5 dari maks 5.0 juga kemungkinan sebagai Spam 49%

Dari Deskripsi data di atas akan di pastikan kembali menggunakan Correlation Heatmap apakah kolom-kolom tersebut sangat berpengaruh pada penentuan `Spam Indicator`

